R/survival_score.R

```diff
@@ -0,0 +1,262 @@
+#' @keywords internal
+h_lr_score_no_strata_no_cov <- function(
+  theta,
+  df,
+  treatment,
+  time,
+  status,
+  n = nrow(df),
+  use_ties_factor = TRUE
+) {
+  assert_numeric(theta, min.len = 1L, finite = TRUE)
+  assert_data_frame(df)
+  assert_string(treatment)
+  assert_string(time)
+  assert_string(status)
+  assert_factor(df[[treatment]], n.levels = 2L, any.missing = FALSE)
+  assert_numeric(df[[status]], any.missing = FALSE)
+  assert_true(all(df[[status]] %in% c(0, 1)))
+  assert_numeric(df[[time]], lower = 0, any.missing = FALSE)
+  assert_count(n)
+  assert_true(n >= nrow(df))
+  assert_flag(use_ties_factor)
+
+  # Standardize data set format, subset to relevant variables.
+  df_stand <- data.frame(
+    treatment_numeric = as.numeric(df[[treatment]]) - 1L,
+    time = df[[time]],
+    status = df[[status]]
+  )
+
+  # Sort by time.
+  df_stand <- df_stand[order(df_stand$time), , drop = FALSE]
+
+  # Patients with events.
+  df_events <- df_stand[df_stand$status == 1L, , drop = FALSE]
+
+  # Add number of events per unique event time.
+  df_events_per_time <- h_n_events_per_time(df_stand, time = "time", status = "status")
+  df_events <- merge(df_events, df_events_per_time, by = "time", all = TRUE)
+
+  # Calculate the log rank statistic u_l and the variance sigma_l2 iteratively.
+  u_l <- sigma_l2 <- 0
+  for (i in seq_len(nrow(df_events))) {
+    # This event time.
+    t_i <- df_events$time[i]
+
+    # Number of overall events at this time.
+    n_events_ti <- df_events$n_events[i]
+
+    # This treatment arm indicator.
+    trt_i <- df_events$treatment_numeric[i]
+
+    # Proportions of patients at risk at time t_i, per arm.
+
+    # Note: Also here we need to use sum divided by n, otherwise
+    # we will get wrong results when there are ties.
+    y_bar_1_ti <- sum(df_stand$treatment_numeric & df_stand$time >= t_i) / n
+    y_bar_0_ti <- sum(!df_stand$treatment_numeric & df_stand$time >= t_i) / n
+
+    # Adjusted proportion of patients at risk in the treatment arm:
+    y_bar_1_ti_coxph <- y_bar_1_ti * exp(theta)
+
+    # Increment u_l.
+    u_l <- u_l + (trt_i - y_bar_1_ti_coxph / (y_bar_1_ti_coxph + y_bar_0_ti))
+
+    # Increment sigma_l2.
+    y_bar_all_ti <- y_bar_1_ti_coxph + y_bar_0_ti
+
+    sigma_l2_factor_ti <- if (!use_ties_factor || n_events_ti == 1) {
+      1
+    } else {
+      (n * y_bar_all_ti - n_events_ti) / (n * y_bar_all_ti - 1)
+    }
+    sigma_l2_summand <- y_bar_1_ti_coxph * y_bar_0_ti * sigma_l2_factor_ti / y_bar_all_ti^2
+    sigma_l2 <- sigma_l2 + sigma_l2_summand
+  }
+  u_l <- u_l / n
+  sigma_l2 <- sigma_l2 / n
+  se_theta_l <- sqrt(1 / (n * sigma_l2))
+  structure(
+    u_l,
+    sigma_l2 = sigma_l2,
+    se_theta_l = se_theta_l,
+    n = n
+  )
+}
+
+#' @describeIn survival_score_functions with strata but without covariates.
+#' @keywords internal
+h_lr_score_strat <- function(theta, df, treatment, time, status, strata, use_ties_factor = TRUE) {
```

Comment:
may I ask if only one strata is allowed?

Reply:
This is a separate discussion because it concerns the general RobinCar2 model formula interface, I deferred this good suggestion to https://github.com/openpharma/RobinCar2/issues/71#issuecomment-3117881708
Currently the user would need to define a single `strata` factor via e.g. `interaction()`